homeassistant/components/decora_wifi/__init__.py

```diff
@@ -1 +1,111 @@
+"""Support for the myLeviton decora_wifi component."""
+
+import logging
+
+import voluptuous as vol
+
+from homeassistant import config_entries
+from homeassistant.const import (
+    CONF_HOST,
+    CONF_PASSWORD,
+    CONF_SCAN_INTERVAL,
+    CONF_USERNAME,
+)
+from homeassistant.core import HomeAssistant
+import homeassistant.helpers.config_validation as cv
+
+from .common import DecoraWifiCommFailed, DecoraWifiPlatform, DecoraWifiSessionNotFound
+from .const import CONF_OPTIONS, DEFAULT_SCAN_INTERVAL, DOMAIN, PLATFORMS
+
+_LOGGER = logging.getLogger(__name__)
+
+DECORAWIFI_HOST_SCHEMA = vol.Schema({vol.Required(CONF_HOST): cv.string})
+
+CONFIG_SCHEMA = vol.Schema(
+    vol.All(
+        cv.deprecated(DOMAIN),
+        {
+            DOMAIN: vol.Schema(
+                {
+                    vol.Required(CONF_USERNAME): cv.string,
+                    vol.Required(CONF_PASSWORD): cv.string,
+                    vol.Optional(CONF_SCAN_INTERVAL, default=120): cv.positive_int,
+                },
+            )
+        },
+    ),
+    extra=vol.ALLOW_EXTRA,
+)
+
+
+async def async_setup(hass, config):
+    """Set up the Decora Wifi component."""
+    hass.data.setdefault(DOMAIN, {})
+
+    if DOMAIN not in config:
+        return True
+
+    # Trigger import config flow
+    conf = config[DOMAIN]
+
+    if not hass.config_entries.async_entries(DOMAIN):
+        hass.async_create_task(
+            hass.config_entries.flow.async_init(
+                DOMAIN,
+                context={"source": config_entries.SOURCE_IMPORT},
+                data={
+                    CONF_USERNAME: conf[CONF_USERNAME],
+                    CONF_PASSWORD: conf[CONF_PASSWORD],
+                },
+            )
+        )
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry):
+    """Set up Decora Wifi from a config entry."""
+
+    conf_data = dict(entry.data)
+    email = conf_data[CONF_USERNAME]
+    password = conf_data[CONF_PASSWORD]
+
+    conf_data[CONF_OPTIONS] = {}
+    # If no scan interval option was set in the config entry, use the default.
+    conf_data[CONF_OPTIONS].update(
+        {
+            CONF_SCAN_INTERVAL: dict(entry.options).get(
+                CONF_SCAN_INTERVAL, DEFAULT_SCAN_INTERVAL
+            )
+        }
+    )
+    devices = {}
+
+    try:
+        devices = await DecoraWifiPlatform.async_getdevices(hass, email)
+    except DecoraWifiCommFailed:
+        _LOGGER.error("Communication with Decora Wifi platform failed.")
+        return False
+    except DecoraWifiSessionNotFound:
+        # Re-login.
+        if await DecoraWifiPlatform.async_login(hass, email, password):
+            devices = await DecoraWifiPlatform.async_getdevices(hass, email)
+
+    # Forward the config entry for each device type present.
+    for p in PLATFORMS:
+        if devices[p]:
+            hass.async_create_task(
+                hass.config_entries.async_forward_entry_setup(entry, p)
+            )
```

Comment:
```suggestion
    hass.config_entries.async_setup_platforms(entry, PLATFORMS)
```

Reply:
The upstream's sync API needs to be wrapped up in async coroutines in order to be called during the config flow (to verify the user-provided credentials are valid, for instance), because config flow is async-only.  Once that work is done, it seems to me like it makes more sense to just continue on implementing the whole integration async, using wrappers around calls to the upstream's sync API as needed.

Specifically regarding the above call to `async_forward_entry_setup', (which the other reviewer commented on as well), I was following the official documentation [here](https://developers.home-assistant.io/docs/config_entries_index/#for-platforms); I haven't been able to locate references to `async_setup_platforms` in the documentation so I'm not sure how to use that. Will research this more.

Reply:
We should avoid wrapping the API in async methods. We have a sync API available on most places (not on the entry, but all other occasions like turn_on/turn_off/update a most of those).

Reply:
Got it.